charmhelpers/contrib/templating/jinja.py

```diff
@@ -31,8 +31,21 @@
 DEFAULT_TEMPLATES_DIR = 'templates'
 
 
-def render(template_name, context, template_dir=DEFAULT_TEMPLATES_DIR):
+def render(template_name, context, template_dir=DEFAULT_TEMPLATES_DIR,
+           jinja_env_args=None):
```

Comment:
Change lgtm. I wonder if `jinja_env_var=None` could be replaced by `jinja_env_args={}` or directly by `**env_kwargs`. On line 47, there is no input validation so the behavior would be the same if the dict key does not exist in `jinja2.Environment`. The way params are passed to this `render(...)` helper would also change.

This is not a blocker to me, though.

Reply:
This would be a really *bad* idea.  Setting the optional parameter to `None` **is** the pattern to indicate that nothing has been passed.

The reason is that Python passes parameters by reference, if if the default for a parameter is a boxed (i.e. a reference to a type) like `Dict` or `List` then *only on the first call* is the default param initialised.  Thus, it acts like a local cache and is the same dictionary on each invocation:

```
$ ipython3
Python 3.8.5 (default, Jan 27 2021, 15:41:15)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.13.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: def a(x, y={}):
   ...:     y[x] = 1
   ...:     print(y)
   ...:

In [2]: a('hello')
{'hello': 1}

In [3]: a('goodbye')
{'hello': 1, 'goodbye': 1}
```

So this is that correct pattern.

Reply:
I was thinking about both of those approaches.

`jinja_env_args={}`. I generally try to avoid using mutable default argument values even in cases like this where it would be theoretically safe. (@ajkavanagh has more thorough explanation, thanks)

`**env_kwargs`. I thought that this could potentially create some confusion because `Environment.render()` (for which this function is basically a wrapper) also takes `**kwargs` but for a different purpose and I can see how someone could think that `**kwargs` in `charmhelpers.render` and `jinja.render` behave the same. I don't know, maybe I'm overthinking this one...

Reply:
(@mkalcok sorry for hijacking your review ...)

Just to follow-up on my comment, the initialization of the defaults happens at *function definition* not at first call, like I erroneously said.  This means that expensive defaults happen when the *module is loaded*:

```python
In [13]: class B(): 
    ...:     def __init__(self): 
    ...:         self.x = 5 
    ...:         print("hi from B() init") 
    ...:          
    ...:                                                                                        

In [14]: def b(x=B()): 
    ...:     print("hi from b() function") 
    ...:                                                                                        
hi from B() init

In [15]: b()                                                                                    
hi from b() function
```

But there's also a really important point here.  As python passes *everything* by reference, any type with interior mutability (e.g. Dict, List, instantiated objects) is vulnerable to both the "default kwarg" problem, but also the "in-out" return problem.  "Simple" types like ints don't have interior mutability and, thus, aren't susceptible.

So the pattern, generally, when accepting an interior-mutable type into a function that that function is going to modify is to make a copy of it, so as not to contaminate the caller's 'version' of the variable.  e.g.

```python
def f(some_list, some_dict):
    my_list = some_list[:]
    my_list.append(something)
    my_dict = some_dict.copy()  # shallow copy only!
    my_dict['extra'] = thing
    # etc.
```

For classes, it's a bit muddier (as there's no default 'copy' for them), but the strategy/pattern should always be "avoid modifying the callers parameters".

Again, sorry for hijacking the thread, but just wanted to be clear.  Thanks!

Reply:
Cristal clear. +1 on the approach.